### DISCUSSION TITLE
Migration Roadmap

### DISCUSSION THREAD

**Discussion by nathan-isaac** [nathanjisaac/actual-sync#1] [General]
We may want to break some of these into other discussions but for now I'm going to consolidate them all here. 

## Contributions

I updated the [readme](https://github.com/nathanjisaac/actual-sync/blob/main/README.md) to include some high level architecture plans.

## Routes to migrate

- [x] `/*` Static Assets

- [x] `/account` 
  - [x] [GET] `/needs-bootstrap`
  - [x] [GET] `/validate`
  - [x] [POST] `/bootstrap`
  - [x] [POST] `/login`
  - [x] [POST] `/change-password`

- [x] `/sync`
  - [x] [POST] `/sync`
  - [x] [POST] `/user-get-key` 
  - [x] [POST] `/user-create-key` 
  - [x] [POST] `/reset-user-file` 
  - [x] [POST] `/upload-user-file` 
  - [x] [POST] `/download-user-file` 
  - [x] [POST] `/update-user-filename` 
  - [x] [POST] `/delete-user-file` 
  - [x] [GET] `/list-user-files`
  - [x] [GET] `/get-user-file-info`

## CRDT

- [x] Migrate [merkle.js](https://github.com/actualbudget/actual/blob/master/packages/loot-core/src/server/merkle.js)
- [x] Migrate [timestamp.js](https://github.com/actualbudget/actual/blob/master/packages/loot-core/src/server/timestamp.js)

## Protobuf

- [x] Migrate protobuf logic - https://github.com/protocolbuffers/protobuf-go

## Other tasks

- [x] Automate test on push
- [ ] Automate releases with https://goreleaser.com/

**Comment by irfanbacker** [comment]
Hey, SantaClaus from discord here.The proposed architecture looks great! But I'm not exactly sure if assigning tasks according to the routes is ideal at this initial stage. All of these routes contain underlying module dependencies which overlap. I think implementing these modules could be done initially and then the routes could be migrated individually.

The modules could be seperated as:
- Timestamp and Merkle (timestamp.js and merkle.js from loot-core/src/server)
- SyncProtobuf
- Database

What do you think?

**Reply by nathan-isaac** [reply]
Hey, @irfanbacker. Yeah, I was thinking about that.

It does look like the account routes are self contained. So those could be done independently and probably should be done first since some of the `app-account.js` functions are used in the `app-sync.js` code. 

For the syncing routes, you are right, we will need to break that up in some way.

I'm thinking for `Protobuf` we will use something like https://github.com/golang/protobuf to handle the serialization and desensitization. I might stick that in the `routes` package since it's responsible for request and response data. 

I'm not 100% sure on the Timestamp and Merkle logic. It seems like core logic so maybe it would go under a `internal.core.crdt` package (which doesn't exist yet)?

Thoughts?

**Reply by irfanbacker** [reply]
The account routes are self-contained but depend upon the database setup.

The syncing routes depend upon both Database as well as the core CRDT logic. As you suggested, creating `internal.core.crdt` for Timestamp and Merkle logic seems to be the right way to go. We could use [https://github.com/roberson-io/mmh3](https://github.com/roberson-io/mmh3) and [https://github.com/google/uuid](https://github.com/google/uuid) as dependencies for Timestamp. The Merkle logic won't need any external dependencies.

The package you mentioned for protobuf is deprecated, I believe. We could use [protobuf-go](https://github.com/protocolbuffers/protobuf-go) to generate it. The `.proto` file is located [here](https://github.com/actualbudget/actual/blob/master/packages/loot-core/src/server/sync/sync.proto).

So, what I meant earlier was that, there are 3 self-contained tasks that we could assign ourselves right now. We could proceed to the other tasks after completing these 3 initiallly. All the other leftover tasks will be self-contained after this.

**Reply by irfanbacker** [reply]
Hey @nathanjisaac . I have implemented the basic database functions in [my fork](https://github.com/irfanbacker/actual-sync/tree/database). Do take a look at it. I have implemented every function for `WrappedDatabase` except for transactions. The API in golang is very different from Javascript for this. So I think it'll be easier to use the default API from [database/sql](https://pkg.go.dev/database/sql) like `db.Begin()`.

Let me know if any changes are needed. I'll create a PR for this if the changes are fine.

**Reply by nathan-isaac** [reply]
Awesome. Thanks. I'll take a look.

**Reply by irfanbacker** [reply]
I have generated the protobuf from `sync.proto`. I have added a readme on how to regenerate it, if needed. Also, I have created PRs for both of these. Ig that'll make it easier for reviewing.

**Reply by nathan-isaac** [reply]
Awesome. Planning to look at this later this afternoon.
